lib/core/text/u16_string.nit

```diff
@@ -0,0 +1,162 @@
+
+# Wrapper of ICU's UTF-16 encoded strings and conversion
+# This module is meant to ease the use of complex string operations provided by the ICU library.
+# The module provides a wrapper for ICU's string structure : `UChar *` as well as conversion functions to/from `String` and `CString`
+module u16_string is pkgconfig ("icu-io", "icu-i18n", "icu-uc")
+
+`{
+	#include <assert.h>
+	#include <unicode/utypes.h>
+	#include <unicode/ustring.h>
+
+	#include <unicode/uiter.h>
+`}
+
+# UTF-16 encoded string
+class U16String
+	super Finalizable
+	super Text
+
+	# Pointer to a `UChar *` string
+	private var uchar_string: nullable UCharString
+
+	# Number of code units `self` has available
+	private var capacity = 0
+
+	# Returns an empty `U16String` of length `len` or a NULL `U16String` if no `len` parameter is provided
+	init (len: nullable Int) do
+		if len == null then
+			uchar_string = new UCharString.nul
+		else
+			assert len >= 0
+			uchar_string = new UCharString.empty(len)
+			capacity = len - 1
+		end
+	end
+
+	# Returns a converted `U16String` from a `String`
+	init from_string(source: String) do
+		var csource = source.to_cstring
+		var csource_length = source.to_cstring.cstring_length
+
+		uchar_string = new UCharString.nul
+		var required_length = uchar_string.from_cstring(0, csource, csource_length)
+
+		uchar_string = new UCharString.empty(required_length + 1)
+		uchar_string.from_cstring(required_length + 1, csource, csource_length)
+		capacity = required_length
+	end
+
+	redef fun length: Int do return uchar_string.length
+
+	# Returns the number of code units in `self`
+	fun nb_code_units: Int do return uchar_string.nb_code_units
+
+	# Copies the characters of `source` to `self`
+	# A maximum of `self.capacity` code units will be copied to `self`
+	fun copy_from(source: String) do
+		uchar_string.from_cstring(capacity,source.to_cstring, source.to_cstring.cstring_length)
+	end
+
+	redef fun chars: Sequence[Char] do
+		var arr = new Array[Char]
+		var offset = 0
+
+		for i in [0..length[ do
+			var c = uchar_string.char_at_offset(offset)
+			arr.push(c)
+			offset += 1
+			if c.to_i > 0xFFFF then offset += 1
+		end
+		return arr
+	end
+
+	redef fun to_cstring: CString do
+		var cself = new CString.nul
+		var required_length = uchar_string.to_cstring(cself, 0, length)
+
+		cself = new CString(required_length + 1)
+		uchar_string.to_cstring(cself, required_length + 1, length)
+
+		return cself
+	end
+
+	redef fun to_s: String do return to_cstring.to_s
+	redef fun finalize do uchar_string.free
+end
+
+# ICU string `UChar *` which are UTF-16 strings
+extern class UCharString `{ UChar *`}
+
+	# Returns an empty `UCharString` of length `length`
+	new empty (length: Int) `{
+		UChar * str = (UChar *)malloc(sizeof(UChar) * length);
+		u_memset(str, 0, length);
+		return str;
+	`}
+
+	# Returns a `NULL` `UCharString`
+	new nul `{ return NULL; `}
+
+	# Returns the number of code points until the termination character `\0`
+	fun length: Int `{
```

Comment:
In order to handle correctly `\0`, I propose you add a `capacity` parameter (in number of code units) instead of asking `u_strlen`.